dart/sign_in_with_google/lib/main.dart

```diff
@@ -0,0 +1,181 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:dart_appwrite/dart_appwrite.dart';
+import 'package:dart_appwrite/models.dart';
+import 'package:http/http.dart' as http;
+import 'package:jose/jose.dart';
+
+Future<dynamic> main(final context) async {
+  final requiredEnvVars = [
+    'GOOGLE_CLIENT_ID',
+    'APPWRITE_FUNCTION_API_ENDPOINT',
+    'APPWRITE_FUNCTION_PROJECT_ID',
+  ];
+  for (var varName in requiredEnvVars) {
+    if (Platform.environment[varName]?.isEmpty ?? true) {
+      throw Exception('Environment variable $varName must be set.');
+    }
+  }
+
+  final googleClientId = Platform.environment['GOOGLE_CLIENT_ID']!;
+
+  final reqBody = context.req.bodyJson as Map<String, dynamic>;
+  final idToken = reqBody['idToken'] ?? '';
+
+  // Validate input
+  if (idToken.isEmpty) {
+    throw Exception('idToken must be provided in the request body.');
+  }
+
+  // Fetch Google's public keys for JWT verification
+  http.Response certsResponse;
+  try {
+    certsResponse = await http
+        .get(
+      Uri.parse('https://www.googleapis.com/oauth2/v3/certs'),
+    )
+        .timeout(
+      const Duration(seconds: 5),
+      onTimeout: () {
+        throw TimeoutException(
+          'Request to fetch Google public keys timed out after 5 seconds',
+        );
+      },
+    );
+  } on TimeoutException catch (e) {
+    context.log('Timeout fetching Google public keys: $e');
+    throw Exception(
+        'Failed to fetch Google public keys: Request timed out. Please try again.');
+  } catch (e) {
+    context.log('Error fetching Google public keys: $e');
+    throw Exception('Failed to fetch Google public keys: $e');
+  }
```

Comment:
Let's also cache the response based on the Cache-Control header as directed by [Google](https://developers.google.com/identity/gsi/web/guides/verify-google-id-token):

> These keys are regularly rotated; examine the Cache-Control header in the response to determine when you should retrieve them again.